viz/RobotModel.cpp

```diff
@@ -174,6 +183,142 @@ void OSGSegment::attachVisual(boost::shared_ptr<urdf::Visual> visual, QDir baseD
+    sdf_pose_to_osg(sdf_visual->GetElement("pose"), *to_visual);
+
+//    toTipOsg_->addChild(to_visual);
+    post_transform_->addChild(to_visual);
+
+    osg::Node* osg_visual = 0;
+    if (sdf_visual->HasElement("geometry")){
+        sdf::ElementPtr sdf_geometry  = sdf_visual->GetElement("geometry");
+        sdf::ElementPtr sdf_geom_elem = sdf_geometry->GetFirstElement();
+
+        if (sdf_geom_elem->GetName() == "box"){
+            osg::Vec3f size;
+            sdf_size_to_osg(sdf_geom_elem->GetElement("size"), size);
+            osg::ShapeDrawable* drawable = new osg::ShapeDrawable(new osg::Box(osg::Vec3(0,0,0), size.x(), size.y(), size.z()));
+            osg_visual = new osg::Geode;
+            osg_visual->asGeode()->addDrawable(drawable);
+        }
+        else if (sdf_geom_elem->GetName() == "cylinder"){
+            double radius = sdf_geom_elem->GetElement("radius")->Get<double>();
+            double length = sdf_geom_elem->GetElement("length")->Get<double>();
+            osg::ShapeDrawable* drawable = new osg::ShapeDrawable(new osg::Cylinder(osg::Vec3d(0,0,0), radius, length));
+            osg_visual = new osg::Geode;
+            osg_visual->asGeode()->addDrawable(drawable);
+        }
+        else if (sdf_geom_elem->GetName() == "sphere"){
+            double radius = sdf_geom_elem->GetElement("radius")->Get<double>();
+            osg::ShapeDrawable* drawable = new osg::ShapeDrawable(new osg::Sphere(osg::Vec3d(0,0,0), radius));
+            osg_visual = new osg::Geode;
+            osg_visual->asGeode()->addDrawable(drawable);
+        }
+        else if  (sdf_geom_elem->GetName() == "mesh"){
+            osg::Vec3 scale;
+            sdf_scale_to_osg(sdf_geom_elem->GetElement("scale"), scale);
+
+            to_visual->setScale(scale);
+
+            std::string uri = sdf_geom_elem->GetElement("uri")->Get<std::string>();
+
+            std::string model_prefix = "model://";
+
+            std::string filename = "";
+
+            if(uri.compare(0, model_prefix.length(), model_prefix) == 0){
+                filename = uri.substr(model_prefix.length());
+            }
+            else {
+                filename = uri;
+            }
+
+            QString qfilename = QString::fromStdString(filename);
+            if (QFileInfo(qfilename).isRelative()){
+                filename = baseDir.absoluteFilePath(QFileInfo(qfilename).fileName()).toStdString();
+//                filename = baseDir.absoluteFilePath(QFileInfo(qfilename).baseName()).toStdString();
+//                filename += ".osg";
+            }
+
+            osg_visual = osgDB::readNodeFile(filename);
+
+            if (!osg_visual) {
+                LOG_ERROR("OpenSceneGraph did not succees loading the mesh file %s.", filename.c_str());
+                throw std::runtime_error("Error loading mesh file.");
+            }
+
+        }
+        else {
+            osg_visual = new osg::Geode;
```

Comment:
At least issue a warning for the user to know that you ignore the defined visual because you can't handle the visual type.